### PR TITLE
Charmhub API find client

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

This is a very mechanical change for adding the charmhub find client
methods to the existing charmhub client.

The find methods takes a new `params.Query` instead of a `params.Entity` as
you might want to utilize search queries that don't align to our strict
names tagging system.

For example, searching for `"word press"` (with a space) will return
`wordpress`, but would fail our names.ApplicationTag parsing. Additionally
searching for multiple things `"wordpress + mysql"` would hopefully be
clever enough to work that out as well.

Essentially we're allowing free form text in the search. We may want to
restrict the amount you can actually search for, but we don't restrict
input anywhere else and to be honest, we'll probably get it wrong.

## QA steps

```sh
go test -v ./api/charmhub/...
```

### CODE REVIEW

**Comment by SimonRichardson** [comment]
@hmlanigan the outstanding issue here is and one we really need to think about, is what happens if you change the charmhub URL `model-config`? I'm unsure it's wise to allow somebody to do this on the fly... let's assume that you do allow the operator to change this, then all unique instance ids for charms/bundles won't match up with the charmhub, potentially breaking an upgrade path. Additionally what happens if the other charmhub doesn't even have the charm/bundle? These seem like dangerous edge cases and very hard to diagnose if somebody does change it?

File: charmhub/transport/find.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ type FindResponse struct {
 	Type           string       `json:"type"`
 	ID             string       `json:"id"`
 	Name           string       `json:"name"`
-	Charm          Charm        `json:"charm,omitempty"`
+	Entity         Entity       `json:"charm"`
```
Comment (hmlanigan) [inline charmhub/transport/find.go:15]: Do we want a TODO for changing the json later, when changed with API?

File: charmhub/info.go [hunk 2]
```diff
@@ -37,5 +37,6 @@ func (c *InfoClient) Info(ctx context.Context, name string) (transport.InfoRespo
 	if err := c.client.Get(ctx, path, &resp); err != nil {
 		return resp, errors.Trace(err)
 	}
+
```
Comment (hmlanigan) [inline charmhub/info.go:40]: Normally a nit, I wouldn't comment on, but seem to be the only change in this file.

File: apiserver/params/charmhub.go [hunk 3]
```diff
@@ -3,20 +3,41 @@
...
+	Results []FindResponse `json:"result"`
+	Errors  ErrorResponse  `json:"errors"`
+}
+
+type FindResponse struct {
+	Type           string         `json:"type"`
+	ID             string         `json:"id"`
+	Name           string         `json:"name"`
+	Entity         CharmHubEntity `json:"entity"`
+	ChannelMap     []ChannelMap   `json:"channel-map"`
```
Comment (hmlanigan) [inline apiserver/params/charmhub.go:39]: Same as for other FindResponse ChannelMaps

File: charmhub/transport/find.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ type FindResponse struct {
...
 	ChannelMap     []ChannelMap `json:"channel-map"`
```
Comment (hmlanigan) [inline charmhub/transport/find.go:16]: ChannelMap is not defined in the Find api docs.

**Comment by hmlanigan** [comment]
@SimonRichardson agreed on the model-config change.  We should write this up.

File: apiserver/params/charmhub.go [hunk 5]
```diff
@@ -3,8 +3,15 @@
 
 package params
 
+// Query holds the query information when attempting to find possible charms or
+// bundles for searching the charmhub.
+type Query struct {
+	Query string `json:"query"`
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/params/charmhub.go:9]: Yeah, I just can't test it, so haven't done anything yet.

File: api/charmhub/data.go [hunk 6]
```diff
@@ -90,6 +112,15 @@ type InfoResponse struct {
 	DefaultRelease ChannelMap   `json:"default-release,omitempty"`
 }
 
+type FindResponse struct {
+	Type           string       `json:"type"`
+	ID             string       `json:"id"`
+	Name           string       `json:"name"`
+	Charm          Charm        `json:"charm"`
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply api/charmhub/data.go]: So this is only optional if you pass in the fields flag. I don't think we want to do that, as we always want all the fields all the time. I see fields being useful for displaying things in the web UI, but for us, we always want all the fields, all the time.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**